3-UsingAPIs/Week1/assignment/ex1-johnWho.js

```
@@ -7,22 +7,28 @@ Rewrite this function, but replace the callback syntax with the Promise syntax:
 - If the Promise `rejects`, pass an error as the argument to reject with: "You 
   didn't pass in a first name!"
 ------------------------------------------------------------------------------*/
-// TODO see above
-export const getAnonName = (firstName, callback) => {
-  setTimeout(() => {
-    if (!firstName) {
-      callback(new Error("You didn't pass in a first name!"));
-      return;
-    }
-
-    const fullName = `${firstName} Doe`;
-
-    callback(fullName);
-  }, 1000);
+export const getAnonName = (firstName) => {
+  return new Promise((resolve, reject) => {
+    setTimeout(() => {
+      if (firstName) {
```

This looks great but a couple of things we can improve are:

1. prevent redundant return statement:
The `return` statement after `resolve(fullName)` is unnecessary because the resolve call ends the execution of the promise immediately.

2. "Fail first" or "Fail Fast" approach:
It is sometimes good to identify and report errors as early as possible in our code or development cycle. The benefit is that, it helps prevent bugs from propagating and becoming more difficult and costly to fix later on.

Original solution:
```
...
setTimeout(() => {
      if (firstName) {
        const fullName = `${firstName} Doe`;
        resolve(fullName);
        return;
      } else {
        reject(new Error("You didn't pass in a first name!"));
      }
    }, 1000);
...
```

Alternative: Here, we handle the case where the promise reject first. In addition, since, we use `return` statement in the `if () {}` statement, we don't have to add `else {}` block to our code. 
```
...
setTimeout(() => {
  if (!firstName) {
    reject(new Error("You didn't pass in a first name!"));
    return;
  }
  resolve(`${firstName} Doe`);
}, 1000);
...
```